Documentation/EtcdWatch.md

```diff
@@ -0,0 +1,110 @@
+2. call [resumeWatchers](#resumewatchers-function) to resume all working watchers.
+
+## cancelWatch function
+
+Cancel the watch task with the watcher, the `onCanceled` will be called after successfully canceled.
+
+1. The watcher will be removed from [watchers](#watchers-instance) map.
+2. If the [watchers](#watchers-instance) map contain the watcher, it will be moved to [cancelWatchers](#cancelwatchers) and send cancel request to [requestStream](#requeststream-instance).
+
+## requestStream instance
+
+StreamObserver instance
+
+1. It is created by gRPC call `watch`.
+1. It will be a single instance and automatically created by [getRequestStream](#getrequeststream-function) if null.
+2. `requestStream` is used to send request to etcd server for watch creation/cancel.
+3. If error, this stream will be canceled by server and we need to resume this stream by set it to `null` and the [getRequestStream](#getrequeststream-function) will create a new one.
+
+## Watcher Class
+
+1. Hold callback for Watcher creation/cancel/resume/event.
+2. Hold WatcherOption and key for resume.
+3. Hold last revision for resume.
+
+## watchers instance
+
+ConcurrentHashMap collection for working watcher.
+
+1. It is used for WatchResponse distribution.
+2. It is used for resumes.
+
+## pendingWatchers instance
+
+It hold the on creating watchers.
+
+## cancelWatchers
+
+ConcurrentHashMap collection for canceling watcher.
+
+1. It is used to filter canceled watch events response.
+2. It is used to hold the canceling watcher.
+3. The watcher will be deleted from `cancelWatchers` after canceled successfully.
+
+## getRequestStream function
+
+1. Single instance method to get [requestStream](#requeststream-instance).
+2. Create requestStream with gRPC call `watch` with responseStream.
+2. The responseStream will distribute the create, cancel, normal response to [processCreate](#processcreate-function), [processCanceled](#processcanceled-function) and [processEvents](#processevents-function).
+3. If error happened, the [requestStream](#requeststream-instance) will be closed by server side, so we call resume to resume all ongoing watchers.
+
+## processCreate function
+
+Process create response from etcd server.
+
+1. If there is no pendingWatcher, ignore.
+2. If cancel flag is true or `CompactRevision!=0`  means the start revision has been compacted out of the store, call onCreateFailed.
+3. If watchId = -1, create failed, call `onCreateFailed`.
+4. If everything is Ok, create watcher, complete ListenableFuture task and put the new watcher to the [watchers](#watchers-instance) map.
+
+## processCanceled function
+
+Process cancel response from etcd server.
+
+1. Remove the respond watcher from [cancelWatchers](#cancelwatchers).
+2. call `onCancel` callback.
+
+
+## resumeWatchers function
+
+Resume all the the watchers on new requestStream.
+
+1. Build new watch creation request for old watcher with last revision.
```

Comment:
@heyitsanthony , I will set the revision = last revision, and filter the events with last revision to clear the resuming status.